File: 73.set-matrix-zeroes.java
/*
 * @lc app=leetcode id=73 lang=java
 *
 * [73] Set Matrix Zeroes
 */

// @lc code=start
class Solution {
    public void setZeroes(int[][] matrix) {
        int rowSize = matrix.length;
        int colSize = matrix[0].length;
        boolean isFirstRowZeros = false;
        boolean isFirstColZeros = false;

        for (int col = 0; col < colSize; col++) {
            if (matrix[0][col] == 0) {
                isFirstRowZeros = true;
                break;
            }
        }

        for (int row = 0; row < rowSize; row++) {
            if (matrix[row][0] == 0) {
                isFirstColZeros = true;
                break;
            }
        }

        for (int row = 1; row < rowSize; row++) {
            for (int col = 1; col < colSize; col++) {
                if (matrix[row][col] == 0) {
                    matrix[0][col] = 0;
                    matrix[row][0] = 0;
                }
            }
        }

        for (int row = 1; row < rowSize; row++) {
            for (int col = 1; col < colSize; col++) {
                if (matrix[0][col] == 0 || matrix[row][0] == 0) {
                    matrix[row][col] = 0;
                }
            }
        }

        if (isFirstColZeros) {
            for (int row = 0; row < rowSize; row++) {
                matrix[row][0] = 0;
            }
        }

        if (isFirstRowZeros) {
            for (int col = 0; col < colSize; col++) {
                matrix[0][col] = 0;
            }
        }

    }
}
// @lc code=end

File: 73.set-matrix-zeroes.py
#
# @lc app=leetcode id=73 lang=python3
#
# [73] Set Matrix Zeroes
#

# @lc code=start
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        
        is_first_row_zeros = False
        is_first_col_zeros = False
        row_size = len(matrix)
        col_size = len(matrix[0])

        for row in range(row_size):
            if matrix[row][0] == 0:
                is_first_col_zeros = True
                break
        
        for col in range(col_size):
            if matrix[0][col] == 0:
                is_first_row_zeros = True
        
        for row in range(1, row_size):
            for col in range(1, col_size):
                if matrix[row][col] == 0:
                    matrix[row][0] = 0
                    matrix[0][col] = 0
        
        for row in range(1, row_size):
            for col in range(1, col_size):
                if matrix[row][0] == 0 or matrix[0][col] == 0:
                    matrix[row][col] = 0
        
        if is_first_col_zeros:
            for row in range(row_size):
                matrix[row][0] = 0
        
        if is_first_row_zeros:
            for col in range(col_size):
                matrix[0][col] = 0
                
            
# @lc code=end

